File: app/src/main/java/net/j4dy/familypicingame/ui/screens/HomeScreen.kt
package net.j4dy.familypicingame.ui.screens

import androidx.compose.animation.core.*
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Face
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import net.j4dy.familypicingame.data.FaceStorage
import net.j4dy.familypicingame.games.GameDescriptor
import net.j4dy.familypicingame.ui.theme.CardSlate
import net.j4dy.familypicingame.ui.theme.CyberPurple
import net.j4dy.familypicingame.ui.theme.DeepDarkBlue
import net.j4dy.familypicingame.ui.theme.ElectricCyan
import net.j4dy.familypicingame.ui.theme.IcyWhite
import net.j4dy.familypicingame.ui.theme.NeonPink
import net.j4dy.familypicingame.ui.theme.SoftGrey
import kotlinx.coroutines.delay

@Composable
fun HomeScreen(
    games: List<GameDescriptor>,
    onGameSelect: (GameDescriptor) -> Unit,
    onManageFacesSelect: () -> Unit
) {
    val context = LocalContext.current
    val faceStorage = remember { FaceStorage(context) }
    var faceCount by remember { mutableStateOf(faceStorage.getProfiles().size) }

    // Update count when navigating back
    LaunchedEffect(Unit) {
        faceCount = faceStorage.getProfiles().size
    }

    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(DeepDarkBlue)
    ) {
        // 1. Beautiful animated cosmic grid / floating particles background
        AnimatedParticleBackground()

        // 2. Main content container
        Column(
            modifier = Modifier
                .fillMaxSize()
                .statusBarsPadding()
                .navigationBarsPadding()
                .padding(24.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.SpaceBetween
        ) {
            // Header
            Column(
                horizontalAlignment = Alignment.CenterHorizontally,
                modifier = Modifier.padding(top = 16.dp)
            ) {
                Text(
                    text = "FAMILY",
                    style = MaterialTheme.typography.labelMedium.copy(
                        fontSize = 14.sp,
                        letterSpacing = 6.sp,
                        color = ElectricCyan
                    ),
                    fontWeight = FontWeight.Black
                )
                Text(
                    text = "Pic-In-Game",
                    style = MaterialTheme.typography.titleLarge.copy(
                        fontSize = 38.sp,
                        lineHeight = 44.sp,
                        fontWeight = FontWeight.Black,
                        brush = Brush.linearGradient(
                            colors = listOf(NeonPink, CyberPurple, ElectricCyan)
                        )
                    )
                )
                Text(
                    text = "Play classic games with family face characters!",
                    style = MaterialTheme.typography.bodyMedium,
                    color = SoftGrey,
                    modifier = Modifier.padding(top = 8.dp),
                    textAlign = TextAlign.Center
                )
            }

            // Games Container
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .weight(1f)
                    .padding(vertical = 32.dp),
                verticalArrangement = Arrangement.Center,
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                Text(
                    text = "SELECT MISSION",
                    color = SoftGrey.copy(alpha = 0.8f),
                    fontSize = 12.sp,
                    fontWeight = FontWeight.Bold,
                    letterSpacing = 2.sp,
                    modifier = Modifier.fillMaxWidth().padding(bottom = 16.dp),
                    textAlign = TextAlign.Start
                )

                games.forEach { game ->
                    GameCard(
                        game = game,
                        onClick = { onGameSelect(game) }
                    )
                    Spacer(modifier = Modifier.height(16.dp))
                }
            }

            // Footer Manager card (Glassmorphic)
            Card(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(90.dp)
                    .border(
                        width = 1.dp,
                        brush = Brush.linearGradient(
                            colors = listOf(ElectricCyan.copy(alpha = 0.5f), Color.Transparent)
                        ),
                        shape = RoundedCornerShape(20.dp)
                    )
                    .clickable { onManageFacesSelect() },
                colors = CardDefaults.cardColors(
                    containerColor = CardSlate.copy(alpha = 0.7f)
                ),
                shape = RoundedCornerShape(20.dp)
            ) {
                Row(
                    modifier = Modifier
                        .fillMaxSize()
                        .padding(horizontal = 20.dp),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.SpaceBetween
                ) {
                    Row(verticalAlignment = Alignment.CenterVertically) {
                        Box(
                            modifier = Modifier
                                .size(50.dp)
                                .clip(RoundedCornerShape(12.dp))
                                .background(NeonPink.copy(alpha = 0.2f)),
                            contentAlignment = Alignment.Center
                        ) {
                            Icon(
                                imageVector = Icons.Default.Face,
                                contentDescription = "Faces",
                                tint = NeonPink,
                                modifier = Modifier.size(30.dp)
                            )
                        }
                        
                        Spacer(modifier = Modifier.width(16.dp))
                        
                        Column {
                            Text(
                                text = "Family Faces",
                                fontWeight = FontWeight.Bold,
                                color = IcyWhite,
                                fontSize = 16.sp
                            )
                            Text(
                                text = "$faceCount Characters Ready",
                                color = SoftGrey,
                                fontSize = 13.sp
                            )
                        }
                    }
                    
                    Text(
                        text = "MANAGE",
                        color = ElectricCyan,
                        fontSize = 13.sp,
                        fontWeight = FontWeight.Bold,
                        letterSpacing = 1.sp
                    )
                }
            }
        }
    }
}

@Composable
fun GameCard(
    game: GameDescriptor,
    onClick: () -> Unit
) {
    val isSlingshot = game.id == "slingshot"
    val borderGradient = Brush.linearGradient(
        colors = if (isSlingshot) {
            listOf(NeonPink, CyberPurple.copy(alpha = 0.2f))
        } else {
            listOf(ElectricCyan, CyberPurple.copy(alpha = 0.2f))
        }
    )

    Card(
        modifier = Modifier
            .fillMaxWidth()
            .height(110.dp)
            .border(width = 1.dp, brush = borderGradient, shape = RoundedCornerShape(24.dp))
            .clickable { onClick() },
        colors = CardDefaults.cardColors(containerColor = CardSlate.copy(alpha = 0.6f)),
        shape = RoundedCornerShape(24.dp)
    ) {
        Row(
            modifier = Modifier
                .fillMaxSize()
                .padding(20.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.SpaceBetween
        ) {
            Column(
                modifier = Modifier.weight(1f),
                verticalArrangement = Arrangement.Center
            ) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Text(
                        text = game.title,
                        fontWeight = FontWeight.Black,
                        fontSize = 20.sp,
                        color = IcyWhite
                    )
                    Spacer(modifier = Modifier.width(8.dp))
                    Text(
                        text = if (isSlingshot) "PHYSICS" else "ARCADE",
                        color = if (isSlingshot) NeonPink else ElectricCyan,
                        fontSize = 10.sp,
                        fontWeight = FontWeight.Bold,
                        modifier = Modifier
                            .background(
                                if (isSlingshot) NeonPink.copy(alpha = 0.15f) else ElectricCyan.copy(alpha = 0.15f),
                                RoundedCornerShape(4.dp)
                            )
                            .padding(horizontal = 6.dp, vertical = 2.dp)
                    )
                }
                
                Spacer(modifier = Modifier.height(4.dp))
                
                Text(
                    text = game.description,
                    color = SoftGrey,
                    fontSize = 13.sp,
                    maxLines = 2
                )
            }
            
            // Visual launcher indicator arrow
            Box(
                modifier = Modifier
                    .size(40.dp)
                    .clip(RoundedCornerShape(12.dp))
                    .background(TranslucentWhite),
                contentAlignment = Alignment.Center
            ) {
                Text(
                    text = "▶",
                    color = IcyWhite,
                    fontSize = 14.sp,
                    fontWeight = FontWeight.Bold
                )
            }
        }
    }
}

private val TranslucentWhite = Color(0x10FFFFFF)

/**
 * Animated Particle Canvas background for a highly premium, futuristic, and responsive feel.
 */
@Composable
fun AnimatedParticleBackground() {
    val infiniteTransition = rememberInfiniteTransition()
    
    // Slow continuous cycle
    val rotation by infiniteTransition.animateFloat(
        initialValue = 0f,
        targetValue = 360f,
        animationSpec = infiniteRepeatable(
            animation = tween(40000, easing = LinearEasing),
            repeatMode = RepeatMode.Restart
        )
    )

    val scaleState by infiniteTransition.animateFloat(
        initialValue = 0.9f,
        targetValue = 1.1f,
        animationSpec = infiniteRepeatable(
            animation = tween(8000, easing = SineIntensityEasing),
            repeatMode = RepeatMode.Reverse
        )
    )

    Canvas(modifier = Modifier.fillMaxSize()) {
        val w = size.width
        val h = size.height
        
        // Draw cosmic glow spheres
        // Sphere 1: Magenta glow top left
        drawCircle(
            brush = Brush.radialGradient(
                colors = listOf(NeonPink.copy(alpha = 0.15f * scaleState), Color.Transparent),
                center = Offset(w * 0.1f, h * 0.2f),
                radius = w * 0.7f
            ),
            radius = w * 0.7f,
            center = Offset(w * 0.1f, h * 0.2f)
        )

        // Sphere 2: Cyan glow bottom right
        drawCircle(
            brush = Brush.radialGradient(
                colors = listOf(ElectricCyan.copy(alpha = 0.12f * scaleState), Color.Transparent),
                center = Offset(w * 0.9f, h * 0.8f),
                radius = w * 0.8f
            ),
            radius = w * 0.8f,
            center = Offset(w * 0.9f, h * 0.8f)
        )
    }
}

private val SineIntensityEasing = Easing { fraction ->
    Math.sin(fraction * Math.PI).toFloat()
}

File: lib/screens/home_screen.dart
import 'dart:math' as math;
import 'dart:ui' as ui;
import 'package:flutter/material.dart';
import '../../theme.dart';
import '../../models/game_descriptor.dart';
import '../../data/face_storage.dart';

class HomeScreen extends StatefulWidget {
  final List<GameDescriptor> games;
  final ValueChanged<GameDescriptor> onGameSelect;
  final VoidCallback onManageFacesSelect;
  final FaceStorage faceStorage;

  const HomeScreen({
    super.key,
    required this.games,
    required this.onGameSelect,
    required this.onManageFacesSelect,
    required this.faceStorage,
  });

  @override
  State<HomeScreen> createState() => _HomeScreenState();
}

class _HomeScreenState extends State<HomeScreen> with SingleTickerProviderStateMixin {
  late AnimationController _bgController;
  int _faceCount = 0;

  @override
  void initState() {
    super.initState();
    _bgController = AnimationController(
      vsync: this,
      duration: const Duration(seconds: 40),
    )..repeat();

    _updateFaceCount();
  }

  void _updateFaceCount() {
    setState(() {
      _faceCount = widget.faceStorage.getProfiles().length;
    });
  }

  @override
  void didUpdateWidget(HomeScreen oldWidget) {
    super.didUpdateWidget(oldWidget);
    _updateFaceCount();
  }

  @override
  void dispose() {
    _bgController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Stack(
        children: [
          // 1. Beautiful animated cosmic background
          AnimatedBuilder(
            animation: _bgController,
            builder: (context, child) {
              return CustomPaint(
                painter: _AnimatedBackgroundPainter(progress: _bgController.value),
                child: Container(),
              );
            },
          ),

          // 2. Main Scrollable Content
          SafeArea(
            child: Padding(
              padding: const EdgeInsets.symmetric(horizontal: 24.0, vertical: 16.0),
              child: Column(
                children: [
                  // Title Header
                  const SizedBox(height: 16),
                  const Text(
                    "FAMILY",
                    style: TextStyle(
                      fontSize: 14,
                      letterSpacing: 6.0,
                      fontWeight: FontWeight.w900,
                      color: electricCyan,
                    ),
                  ),
                  ShaderMask(
                    shaderCallback: (bounds) => const LinearGradient(
                      colors: [neonPink, cyberPurple, electricCyan],
                    ).createShader(bounds),
                    child: const Text(
                      "Pic-In-Game",
                      style: TextStyle(
                        fontSize: 38,
                        fontWeight: FontWeight.w900,
                        color: Colors.white,
                        height: 1.1,
                      ),
                    ),
                  ),
                  const SizedBox(height: 8),
                  const Text(
                    "Play classic games with family face characters!",
                    style: TextStyle(color: softGrey, fontSize: 13),
                    textAlign: TextAlign.center,
                  ),

                  // Games List Section
                  Expanded(
                    child: ListView.separated(
                      padding: const EdgeInsets.symmetric(vertical: 32.0),
                      itemCount: widget.games.length + 1,
                      separatorBuilder: (context, index) => const SizedBox(height: 16),
                      itemBuilder: (context, index) {
                        if (index == 0) {
                          return const Padding(
                            padding: EdgeInsets.only(bottom: 8.0),
                            child: Text(
                              "SELECT MISSION",
                              style: TextStyle(
                                color: Color(0xCC8F9CAE),
                                fontSize: 12,
                                fontWeight: FontWeight.bold,
                                letterSpacing: 2.0,
                              ),
                            ),
                          );
                        }
                        final game = widget.games[index - 1];
                        return _GameCard(
                          game: game,
                          onClick: () => widget.onGameSelect(game),
                        );
                      },
                    ),
                  ),

                  // Footer Manage Faces (Glassmorphic)
                  InkWell(
                    onTap: () async {
                      widget.onManageFacesSelect();
                      // Wait for returning and update count
                      await Future.delayed(const Duration(milliseconds: 300));
                      _updateFaceCount();
                    },
                    borderRadius: BorderRadius.circular(20),
                    child: Container(
                      height: 90,
                      decoration: BoxDecoration(
                        color: cardSlate.withOpacity(0.7),
                        borderRadius: BorderRadius.circular(20),
                        border: Border.all(
                          color: electricCyan.withOpacity(0.3),
                          width: 1.0,
                        ),
                      ),
                      padding: const EdgeInsets.symmetric(horizontal: 20),
                      child: Row(
                        mainAxisAlignment: MainAxisAlignment.spaceBetween,
                        children: [
                          Row(
                            children: [
                              Container(
                                width: 50,
                                height: 50,
                                decoration: BoxDecoration(
                                  color: neonPink.withOpacity(0.2),
                                  borderRadius: BorderRadius.circular(12),
                                ),
                                child: const Icon(
                                  Icons.face_retouching_natural,
                                  color: neonPink,
                                  size: 28,
                                ),
                              ),
                              const SizedBox(width: 16),
                              Column(
                                mainAxisAlignment: MainAxisAlignment.center,
                                crossAxisAlignment: CrossAxisAlignment.start,
                                children: [
                                  const Text(
                                    "Family Faces",
                                    style: TextStyle(
                                      fontWeight: FontWeight.bold,
                                      color: icyWhite,
                                      fontSize: 16,
                                    ),
                                  ),
                                  Text(
                                    "$_faceCount Characters Ready",
                                    style: const TextStyle(
                                      color: softGrey,
                                      fontSize: 13,
                                    ),
                                  ),
                                ],
                              ),
                            ],
                          ),
                          const Text(
                            "MANAGE",
                            style: TextStyle(
                              color: electricCyan,
                              fontSize: 13,
                              fontWeight: FontWeight.bold,
                              letterSpacing: 1.0,
                            ),
                          ),
                        ],
                      ),
                    ),
                  ),
                  const SizedBox(height: 8),
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }
}

class _GameCard extends StatelessWidget {
  final GameDescriptor game;
  final VoidCallback onClick;

  const _GameCard({
    required this.game,
    required this.onClick,
  });

  @override
  Widget build(BuildContext context) {
    final isSlingshot = game.id == 'slingshot';

    return InkWell(
      onTap: onClick,
      borderRadius: BorderRadius.circular(24),
      child: Container(
        height: 110,
        decoration: BoxDecoration(
          color: cardSlate.withOpacity(0.6),
          borderRadius: BorderRadius.circular(24),
          border: Border.all(
            color: isSlingshot ? neonPink.withOpacity(0.8) : electricCyan.withOpacity(0.8),
            width: 1.0,
          ),
        ),
        padding: const EdgeInsets.all(20),
        child: Row(
          children: [
            Expanded(
              child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Row(
                    children: [
                      Text(
                        game.title,
                        style: const TextStyle(
                          fontWeight: FontWeight.w900,
                          fontSize: 18,
                          color: icyWhite,
                        ),
                      ),
                      const SizedBox(width: 8),
                      Container(
                        decoration: BoxDecoration(
                          color: isSlingshot ? neonPink.withOpacity(0.15) : electricCyan.withOpacity(0.15),
                          borderRadius: BorderRadius.circular(4),
                        ),
                        padding: const EdgeInsets.symmetric(horizontal: 6, vertical: 2),
                        child: Text(
                          isSlingshot ? "PHYSICS" : "ARCADE",
                          style: TextStyle(
                            color: isSlingshot ? neonPink : electricCyan,
                            fontSize: 10,
                            fontWeight: FontWeight.bold,
                          ),
                        ),
                      ),
                    ],
                  ),
                  const SizedBox(height: 4),
                  Text(
                    game.description,
                    style: const TextStyle(
                      color: softGrey,
                      fontSize: 12,
                    ),
                    maxLines: 2,
                    overflow: TextOverflow.ellipsis,
                  ),
                ],
              ),
            ),
            const SizedBox(width: 8),
            Container(
              width: 40,
              height: 40,
              decoration: BoxDecoration(
                color: Colors.white.withOpacity(0.06),
                borderRadius: BorderRadius.circular(12),
              ),
              alignment: Alignment.center,
              child: const Icon(
                Icons.play_arrow,
                color: icyWhite,
                size: 20,
              ),
            ),
          ],
        ),
      ),
    );
  }
}

class _AnimatedBackgroundPainter extends CustomPainter {
  final double progress;

  _AnimatedBackgroundPainter({required this.progress});

  @override
  void paint(Canvas canvas, Size size) {
    final w = size.width;
    final h = size.height;

    // Pulse size scale
    final pulseScale = 0.9 + 0.2 * math.sin(progress * 2.0 * math.pi);

    // Draw cosmic glow spheres
    // Sphere 1: Magenta glow top left
    final paint1 = Paint()
      ..shader = ui.Gradient.radial(
        Offset(w * 0.1, h * 0.2),
        w * 0.7,
        [neonPink.withOpacity(0.15 * pulseScale), Colors.transparent],
      );
    canvas.drawCircle(Offset(w * 0.1, h * 0.2), w * 0.7, paint1);

    // Sphere 2: Cyan glow bottom right
    final paint2 = Paint()
      ..shader = ui.Gradient.radial(
        Offset(w * 0.9, h * 0.8),
        w * 0.8,
        [electricCyan.withOpacity(0.12 * pulseScale), Colors.transparent],
      );
    canvas.drawCircle(Offset(w * 0.9, h * 0.8), w * 0.8, paint2);
  }

  @override
  bool shouldRepaint(covariant _AnimatedBackgroundPainter oldDelegate) {
    return oldDelegate.progress != progress;
  }
}
